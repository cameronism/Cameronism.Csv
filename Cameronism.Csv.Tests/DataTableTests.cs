using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text;

namespace Cameronism.Csv.Tests
{
    [TestFixture]
    public class DataTableTests
    {
        private static void Run<T>(IEnumerable<T> items)
        {
            var parameters = typeof(T)
                .GetConstructors()
                .Select(c => c.GetParameters())
                .OrderByDescending(p => p.Length)
                .FirstOrDefault();

            var props = new MethodInfo[parameters.Length];
            var dt = new DataTable();

            for (var i = 0; i < parameters.Length; i++)
            {
                var param = parameters[i];
                dt.Columns.Add(param.Name, Nullable.GetUnderlyingType(param.ParameterType) ?? param.ParameterType);
                props[i] = typeof(T).GetProperty(param.Name).GetGetMethod();
            }


            foreach (var item in items)
            {
                var row = dt.NewRow();
                for (var i = 0; i < props.Length; i++)
                {
                    var value = props[i].Invoke(item, null);
                    row[i] = value ?? DBNull.Value;
                }
                dt.Rows.Add(row);
            }

            var sw = new StringWriter { NewLine = "\r\n" };
            Serializer.Serialize(sw, dt);
            Approvals.VerifyCsv(sw);
        }

        [Test]
        public void OneString()
        {
            //ExecuteReader("select 'ba\nr\"' as foo, 1 as one, null as n");
            Run(Enumerable.Range(1, 10)
                .Select(i => new
                {
                    i,
                    foo = i.ToString(),
                    ix = i % 2 == 0 ? (int?)i : null,
                    dt = new DateTime(2018, 1, i, 0, 0, 0, DateTimeKind.Utc),
                }));
        }
    }
}
