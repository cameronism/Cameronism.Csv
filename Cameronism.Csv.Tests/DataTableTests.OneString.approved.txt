i,foo,ix,dt
1,1,,2018-01-01 00:00:00.000000
2,2,2,2018-01-02 00:00:00.000000
3,3,,2018-01-03 00:00:00.000000
4,4,4,2018-01-04 00:00:00.000000
5,5,,2018-01-05 00:00:00.000000
6,6,6,2018-01-06 00:00:00.000000
7,7,,2018-01-07 00:00:00.000000
8,8,8,2018-01-08 00:00:00.000000
9,9,,2018-01-09 00:00:00.000000
10,10,10,2018-01-10 00:00:00.000000
